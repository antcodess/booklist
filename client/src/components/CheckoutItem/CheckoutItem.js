import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import { addItem, removeItem, clearItem } from '../../redux/cart/cart.actions';

import './CheckoutItem.scss';

const CheckoutItem = ({ item, addItem, removeItem, clearItem, history }) => {
  const { imageUrl, title, quantity, price, routeName } = item;
  return (
    <div className='checkout-item'>
      <div className='image-container' onClick={() => history.push(`/${routeName}`)}>
        <img src={imageUrl} alt='book' />
      </div>
      <span className='title' onClick={() => history.push(`/${routeName}`)}>{title}</span>
      <span className='quantity'>
        <div className='arrow' onClick={() => removeItem(item)}>&#10094;</div>
        <span className='value'>{quantity}</span>
        <div className='arrow' onClick={() => addItem(item)}>&#10095;</div>
      </span>
      <span className='price'>${price}</span>
      <div className='remove' onClick={() => clearItem(item)}>&#10005;</div>
    </div>
  );
};

const mapDispatchToProps = dispatch => ({
  addItem: item => dispatch(addItem(item)),
  removeItem: item => dispatch(removeItem(item)),
  clearItem: item => dispatch(clearItem(item))
});

export default connect(null, mapDispatchToProps)(withRouter(CheckoutItem));