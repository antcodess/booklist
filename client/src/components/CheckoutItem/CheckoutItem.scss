.checkout-item {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr 1fr;
  grid-gap: 10px;
  margin-top: 40px;

  .image-container {
    cursor: pointer;

    img {
      width: 100%;
      height: auto;
    }
  }
  .title,
  .quantity,
  .price,
  .remove {
    margin: auto 0;
  }

  .title {
    cursor: pointer;
  }

  .quantity {
    display: flex;
    justify-content: center;

    .arrow {
      cursor: pointer;
    }

    .value {
      margin: 0 10px;
    }
  }

  .remove {
    cursor: pointer;
  }
}