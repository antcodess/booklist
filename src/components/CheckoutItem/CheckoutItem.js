import React from 'react';
import { connect } from 'react-redux';
import { addItem, removeItem, clearItem } from '../../redux/cart/cart.actions';
import './CheckoutItem.scss';

const CheckoutItem = ({ item, addItem, removeItem, clearItem }) => {
  const { imageUrl, title, quantity, price } = item;
  return (
    <div className='checkout-item'>
      <div className='image-container'>
        <img src={imageUrl} alt='book' />
      </div>
      <span className='title'>{title}</span>
      <span className='quantity'>
        <div className='arrow' onClick={() => removeItem(item)}>&#10094;</div>
        <span className='value'>{quantity}</span>
        <div className='arrow' onClick={() => addItem(item)}>&#10095;</div>
      </span>
      <span className='price'>${price}</span>
      <div className='remove' onClick={() => clearItem(item)}>&#10005;</div>
    </div>
  );
};

const mapDispatchToProps = dispatch => ({
  addItem: item => dispatch(addItem(item)),
  removeItem: item => dispatch(removeItem(item)),
  clearItem: item => dispatch(clearItem(item))
});

export default connect(null, mapDispatchToProps)(CheckoutItem);