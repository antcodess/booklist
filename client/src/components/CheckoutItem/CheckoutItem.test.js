import React from 'react';
import { shallow } from 'enzyme';
import { CheckoutItem } from './CheckoutItem';
import { mapDispatchToProps } from './CheckoutItem';

describe('CheckoutItem Component', () => {
  let wrapper;
  let mockItem;
  let mockAddItem;
  let mockRemoveItem;
  let mockClearItem;
  let mockHistory;

  const mockAddItemAction = {
    type: 'ADD_ITEM', 
    payload: {
      title: 'Item 1', 
      imageUrl: 'example.com', 
      quantity: 1, 
      price: 20, 
      routeName: '/products/item-1'
    }
  }

  const mockRemoveItemAction = {
    type: 'REMOVE_ITEM', 
    payload: {
      title: 'Item 1', 
      imageUrl: 'example.com', 
      quantity: 1, 
      price: 20, 
      routeName: '/products/item-1'
    }
  }

  const mockClearItemAction = {
    type: 'CLEAR_ITEM', 
    payload: {
      title: 'Item 1', 
      imageUrl: 'example.com', 
      quantity: 1, 
      price: 20, 
      routeName: '/products/item-1'
    }
  }
  
  beforeEach(() => {

    mockItem = {
      title: 'Item 1', 
      imageUrl: 'example.com', 
      quantity: 1, 
      price: 20, 
      routeName: '/products/item-1'
    };

    mockAddItem = jest.fn();
    mockRemoveItem = jest.fn();
    mockClearItem = jest.fn();
    mockHistory = { push: jest.fn() };

    const mockProps = {
      item: mockItem, 
      addItem: mockAddItem, 
      removeItem: mockRemoveItem, 
      clearItem: mockClearItem, 
      history: mockHistory
    };

    wrapper = shallow(<CheckoutItem { ...mockProps } />);
  });

  it('should render CheckoutItem Component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call history.push when clicked', () => {
    wrapper.find('.image-container').simulate('click');
    expect(mockHistory.push).toHaveBeenCalled();
  });

  it('should call history.push when clicked 2', () => {
    wrapper.find('.title').simulate('click');
    expect(mockHistory.push).toHaveBeenCalled();
  });

  it('should call removeItem when clicked', () => {
    wrapper.find('#checkout-remove-item').simulate('click');
    expect(mockRemoveItem).toHaveBeenCalled();
  });

  it('should call addItem when clicked', () => {
    wrapper.find('#checkout-add-item').simulate('click');
    expect(mockAddItem).toHaveBeenCalled();
  });

  it('should call clearItem when clicked', () => {
    wrapper.find('.remove').simulate('click');
    expect(mockClearItem).toHaveBeenCalled();
  });

  it('should dispatch addItem action when clicked', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).addItem(mockItem);
    expect(dispatch.mock.calls[0][0]).toEqual(mockAddItemAction);
  });

  it('should dispatch removeItem action when clicked', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).removeItem(mockItem);
    expect(dispatch.mock.calls[0][0]).toEqual(mockRemoveItemAction);
  });
  
  it('should dispatch removeItem action when clicked', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).clearItem(mockItem);
    expect(dispatch.mock.calls[0][0]).toEqual(mockClearItemAction);
  });
});