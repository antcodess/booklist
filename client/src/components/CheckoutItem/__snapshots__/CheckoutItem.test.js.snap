// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutItem Component should render CheckoutItem Component 1`] = `ShallowWrapper {}`;
